### PR TITLE
Np 1990 extend search data

### PR DESCRIPTION
**Pull request by svenngjeto**

### CODE REVIEW

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 1]
```diff
@@ -214,6 +240,47 @@ private static Instant extractPublishedDate(JsonNode record, UUID id) {
         return getInstant(record, id, PUBLISHED_DATE_JSON_POINTER, PUBLISHED_DATE);
     }
 
+    private static Map<String, String> extractAlternativeTitles(JsonNode record, UUID id) {
+        JsonNode node = record.at(ALTERNATIVETITLES_JSON_POINTER);
+        String json = node.toString();
+        Map<String, String> map = Collections.EMPTY_MAP;
```
Comment (brinxmat) [inline search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java:246]: Make this Collections.emptyMap();

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 2]
```diff
@@ -214,6 +240,47 @@ private static Instant extractPublishedDate(JsonNode record, UUID id) {
...
+        try {
+            Map<String, AttributeValue> attributeValueMap = mapper.readValue(json, MAP_ATTRIBUTEVALUE_JAVA_TYPE);
```
Comment (brinxmat) [inline search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java:248]: json, replace with record.at(ALTERNATIVETITLES_JSON_POINTER).toString()

Comment (brinxmat) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java:248]: In fact, extract a method that does this and takes the pointer as a value. Use that everywhere

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 3]
```diff
@@ -214,6 +240,47 @@ private static Instant extractPublishedDate(JsonNode record, UUID id) {
...
+    }
+
+    private static Reference extractDescriptionReference(JsonNode record, UUID id) {
+        JsonNode node = record.at(REFERENCE_JSON_POINTER);
+        String json = node.toString();
+        Reference reference = null;
+        try {
+            AttributeValue referenceAsAttributeValue = mapper.readValue(json, AttributeValue.class);
+            var value = DynamodbItemUtilsClone.toSimpleValue(referenceAsAttributeValue);
+            String j2 = mapper.writeValueAsString(value);
```
Comment (brinxmat) [inline search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java:276]: What is j2?

File: search-commons/src/main/java/no/unit/nva/utils/DynamodbItemUtilsClone.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    private static final ItemValueConformer valueConformer = new ItemValueConformer();
+    public static final String ATTRIBUTE_VALUE_MUST_NOT_BE_EMPTY_MESSAGE = "Attribute value must not be empty: {}";
+
+    @JacocoGenerated
+    private DynamodbItemUtilsClone() {
+    }
+
+
+    /**
+     * Creates an map of attributevalues from json source.
```
Comment (brinxmat) [inline search-commons/src/main/java/no/unit/nva/utils/DynamodbItemUtilsClone.java:49]: Capitalize JSON thoughout

File: search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    @Test
+    void getSampleItemAndGenerateIndexDocumentActuallyProducesIndexDocument() throws IOException {
+        Item item = loadItemFromResourceFile();
+        Assertions.assertNotNull(item);
+        IndexDocument indexDocument =
+                IndexDocumentGenerator.fromJsonNode(objectMapper.valueToTree(ItemUtils.toAttributeValues(item)));
+        Assertions.assertNotNull(indexDocument);
+    }
+
+    private Publication toPublication(String json) {
```
Comment (brinxmat) [inline search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java:200]: Move to bottom

File: search-index-import-util/src/main/java/no/unit/nva/publication/DataPipelineFileReaderPublication.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * @param s3Object BufferedReader containing json dynamodb records
+     */
+    @JacocoGenerated
+    public Long readJsonDataFile(S3Object s3Object) {
+        long processedPublicationCount = 0;
+
+        logger.debug(READING_FROM_S3_MESSAGE, s3Object.getBucketName(), s3Object.getKey());
+        try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(s3Object.getObjectContent()))) {
+            processedPublicationCount = bufferedReader
```
Comment (brinxmat) [inline search-index-import-util/src/main/java/no/unit/nva/publication/DataPipelineFileReaderPublication.java:51]: Can you replace the "bOOL" here?

File: search-testutils/build.gradle [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    implementation group: 'org.mockito', name: 'mockito-core', version: '3.5.7'
+    implementation group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
+    implementation group: 'org.junit.jupiter', name: 'junit-jupiter-api', version: junit5Version
+    implementation group: 'org.junit.jupiter', name: 'junit-jupiter-params', version: junit5Version
+    implementation group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
+
+
+}
+
+//dependencies {
```
Comment (brinxmat) [inline search-testutils/build.gradle:16]: Why is this here?

File: search-testutils/src/main/java/no/unit/nva/utils/TestDataGenerator.java [hunk 8]
```diff
@@ -158,32 +201,57 @@ public IndexDocument asIndexDocument() {
...
+        var rawjson = IoUtils.streamToString(IoUtils.inputStreamFromResources(SAMPLE_DATAPIPELINE_OUTPUT_FILE));
+        return dynamodbExportFormatToItem(rawjson);
+    }
 
     private DynamodbEvent toDynamodbEvent(JsonNode event) {
         return mapper.convertValue(event, DynamodbEvent.class);
     }
 
+    /**
+     * Creates an Item from json source.
```
Comment (brinxmat) [inline search-testutils/src/main/java/no/unit/nva/utils/TestDataGenerator.java:243]: JSON

File: search-testutils/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 9]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (brinxmat) [inline search-testutils/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Is this used?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class DataPipelineFileReaderPublicationTest {
```
Comment (axthosarouris) [inline search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java:31]: DataPipelineFileReaderPublicationTest returns something but there is no test for the happy scenario.

Write the tests for what you want DataPipelineFileReaderPublication without thinking what it already does and how it does it.

File: search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+    private static final String SAMPLE_BUCKET_NAME = "nva-datapipeline";
+    private static final String SAMPLE_S3FOLDER_KEY = "2020-10-12-06-55-32";
+    private static final String SAMPLE_DATAPIPELINE_OUTPUT_FILE = "datapipeline_output_sample";
+    private static final String DUMMY_FILE_NAME = "some_random_file_name";
+    public static final long THE_ANSWER_TO_EVERYTHING = 42L;
```
Comment (axthosarouris) [inline search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java:37]: rename constant

**Review by axthosarouris — CHANGES_REQUESTED** [review]
General comment: Go to the basic modules (mainly handlers)  and test the good and the bad scenarios. E.g. 
handler persists publication when input has no empty attribute values
handler persists publication when input has  empty attribute values
handler logs error when input is not a valid publication
etc.

**Comment by axthosarouris** [comment]
I think what we should be reading from the S3 bucket is not a DAO (DynamoDBStream) format but a DTO. We already have a transformer of a DAO to DTO that publishes an event with a Publication. We want a listener to that event that will write to S3. Then a listener for an S3 event that will take the Publication records and send them to ES

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 12]
```diff
@@ -100,7 +121,12 @@ public static IndexDocumentGenerator fromJsonNode(JsonNode record) {
                 .withAbstract(extractAbstract(record, id))
                 .withPublisher(extractPublisher(record))
                 .withModifiedDate(extractModifiedDate(record, id))
-                .withPublishedDate(extractPublishedDate(record, id));
+                .withPublishedDate(extractPublishedDate(record, id))
+                .withAlternativeTitles(extractAlternativeTitles(record, id))
+                .withTags(extractTags(record,id))
+                .withReference(extractDescriptionReference(record,id))
+                ;
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java]: semi moved

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 13]
```diff
@@ -47,6 +51,9 @@
     public static final String PUBLISHER_TYPE_JSON_POINTER = "/publisher/m/type/s";
     public static final String MODIFIED_DATE_JSON_POINTER = "/modifiedDate/s";
     public static final String PUBLISHED_DATE_JSON_POINTER = "/publishedDate/s";
+    public static final String ALTERNATIVETITLES_JSON_POINTER = "/entityDescription/m/alternativeTitles/m";
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java]: Renamed

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 14]
```diff
@@ -63,6 +70,20 @@
...
     private static final ObjectMapper mapper = JsonUtils.objectMapper;
     private static final Logger logger = LoggerFactory.getLogger(IndexDocumentGenerator.class);
+    public static final String JSON_PROCESSING_EXCEPTION_ON_FIELD_ALTERNATIVE_TITLES =
+            "JsonProcessingException on field 'AlternativeTitles' in record with id={}";
+
+    public static final String JSON_PROCESSING_EXCEPTION_ON_FIELD_REFERENCE =
+            "JsonProcessingException on field 'reference' in record with id={}";
+
+
+    public static final JavaType MAP_ATTRIBUTEVALUE_JAVA_TYPE =
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java]: renamed

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 15]
```diff
@@ -63,6 +70,20 @@
...
+            mapper.getTypeFactory().constructParametricType(Map.class,
+            String.class,
+            AttributeValue.class);
+
+    public static final JavaType LIST_ATTRIBUTEVALUE_JAVA_TYPE =
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java]: renamed

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java (hunk 1 above)
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java]: Done

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java (hunk 3 above)
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java]: ..is a jsonString

File: search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java (hunk 5 above)
Comment (svenngjeto) [reply search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java:200]: Helper methods moved to bottom

File: search-testutils/build.gradle (hunk 7 above)
Comment (svenngjeto) [reply search-testutils/build.gradle]: ..don't copy code, and leave it... Gone!

File: search-testutils/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 9 above)
Comment (svenngjeto) [reply search-testutils/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker]: Nope

File: search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java [hunk 16]
```diff
@@ -214,6 +240,47 @@ private static Instant extractPublishedDate(JsonNode record, UUID id) {
         return getInstant(record, id, PUBLISHED_DATE_JSON_POINTER, PUBLISHED_DATE);
     }
 
+    private static Map<String, String> extractAlternativeTitles(JsonNode record, UUID id) {
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/search/IndexDocumentGenerator.java:237]: The language tag is still the key to the map, maybe it should have been treated as an language attribute into elasticsearch?

File: search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java (hunk 10 above)
Comment (svenngjeto) [reply search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java]: removed DataPipelineFileReaderPublication class completely, was a leftover to be used for datamigration

File: search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java (hunk 11 above)
Comment (svenngjeto) [reply search-index-import-util/src/test/java/no/unit/nva/publication/DataPipelineFileReaderPublicationTest.java]: removed class...

File: search-commons/src/main/java/no/unit/nva/utils/DynamodbItemUtilsClone.java [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+     * @param dynamoDbJson  json source containing publication
+     * @return map of attributevalues created from json source
+     * @throws JsonProcessingException when source contains errors
+     */
+    public static Map<String, AttributeValue> attributeMapFromDynamoDBSource(String dynamoDbJson)
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/utils/DynamodbItemUtilsClone.java]: Thas is now an option...

File: search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@JacocoGenerated
+public class DynamodbItemUtilsCloneTest {
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java:16]: Removed a lot og BS tests, no need to test AWS classes and the tests was just for checking reading from datapipeline files

File: search-commons/src/main/java/no/unit/nva/utils/DynamodbItemUtilsClone.java [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+            throws JsonProcessingException {
+        return objectMapper.readValue(dynamoDbJson, PARAMETRIC_TYPE);
+    }
+
+    /**
+     * Fixes problem with boolean values in source.
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply search-commons/src/main/java/no/unit/nva/utils/DynamodbItemUtilsClone.java]: Boolean values did not read to well from datapipeline output files

File: search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+    private static final String ACTUAL_DATAPIPELINE_OUTPUT_FILE = "31184c66-88a6-47f4-86a5-2334a05d87b2";
+    private static final int EXPECTED_NUMBER_OF_PUBLICATIONS = 32;
+    private static final String SERIALIZED_BOOLEAN_TYPE_TAG = "bOOL";
+    private static final String ATTRIBUTE_VALUE_BOOLEAN_TYPE_TAG = "bool";
+
+    @Test
+    void readingFromDatapipelineFileUsingToPublication() throws IOException {
```
Comment (axthosarouris) [inline search-commons/src/test/java/no/unit/nva/utils/DynamodbItemUtilsCloneTest.java:24]: General comment: The reader of this test is reading "Test reading from datapipeline file using toPublication()"
This is the action. The test here is that DynamodbItemUtilsClone returns Publication objects when the input is a DynamoDB json string. 
In addition the tests do not show why this class is necessary, there should be a test with a single line that has null values and can be parsed.

**Review by axthosarouris — APPROVED** [review]

**Review by brinxmat — APPROVED** [review]
